### ISSUE TITLE
experiment with shipping type annotations

### ISSUE THREAD

**Issue by mmerickel** [Pylons/pyramid#2638]
It'd be cool if pyramid could ship type annotations for its public APIs. I talked to the pycharm folks at their booth last week and there are apparently 3 ways to do this given that we want to keep working on py2 and py3:
1. Use the comment-based type annotation syntax in pyramid source which is compatible with py2.
2. Use separate stub files that are shipped to the typeshed repo. https://github.com/python/typeshed/
3. Use separate stub files distributed in the pyramid package.

It would be helpful to evaluate the pros and cons of the various approaches with the primary goal being to minimize the maintenance effort required. For example, a build hook that could check that type hints were defined for public APIs.

On a side note, I think the commant syntax is ugly and that it should be considered the worst of the 3 options.

Related https://github.com/Pylons/pyramid/issues/2387

NOTE: This issue is vague because I only sort of know what I'm talking about.

**Comment by stevepiercy** [comment]
Option 1 and 3 require a release of Pyramid (con), whereas 2 does not (pro). That alone makes option 2 the most appealing to me.

**Comment by mmerickel** [comment]
> Option 1 and 3 require a release of Pyramid (con), whereas 2 does not (pro). That alone makes option 2 the most appealing to me.

These are type annotations, not documentation. They are inherently coupled to the source of the release. I think this is not a concern.

**Comment by stevepiercy** [comment]
Are you saying that there shall be no bugs or other mistakes in the type annotations when added to the Pyramid package, or that if there are then it is no big thing?

**Comment by mmerickel** [comment]
I think if there are bugs when them then they could simply get fixed in a later release. Again that's just my first impression as I have not used type annotations myself.

**Comment by tseaver** [comment]
-1 to option 2:  annotations have to stay tightly in-sync with code:  in fact, we should find a way to "test" them so that we don't see them drift.  I see no benefit in creating an out-of-tree "authority" for them which can drift (and will, inevitably) from changes as Pyramid releases are made.

**Comment by iivmok** [comment]
I would love to see Pyramid type annotations, personally I like the 3rd option the most. I use PyCharm and type annotations a lot, and marking `request` as WebOb `Request` class helps a whole lot, especially when I first started using Pyramid.

**Comment by ztane** [comment]
I would like to see the option 4: stop supporting Python 2 and use real annotations.

Though a bigger problem is that nothing supports *interfaces*.

**Comment by mmerickel** [comment]
> Though a bigger problem is that nothing supports interfaces.

Yeah `zope.interface` is a big issue for us in Pyramid. It looks like https://github.com/python/mypy/issues/1240 is slowly working on this.

**Comment by enkidulan** [comment]
Until mypy supports `zope.interface` it doesn't seem that it will be much of use of having it at all. With having `typing` around there is no other choice but to steal best approaches from languages that have typing for ages, like `java`, and one of java's good practices is to [use interfaces instead of classes for declaring parameters, return values, etc..](http://jtechies.blogspot.com/2012/07/item-52-refer-to-objects-by-their.html).

**Comment by domenkozar** [comment]
Most of pyramid "exposed" api isn't `zope.interface` and thus would benefit from having types :)

**Comment by mmerickel** [comment]
I'm not sure that's true. We have two major problems in terms of typing / linting.

1. Almost all of our APIs are defined in terms of zope interfaces such as a view signature which accepts `(context: Any, request: pyramid.interfaces.IRequest)` or `config.set_authentication_policy(policy: pyramid.interfaces.IAuthenticationPolicy)`.

2. Some of Pyramid's public api is configurable at runtime as well which is difficult to type at all (speaking as someone with minimal mypy experience). For example `config.add_directive` adds extra methods to the configurator and `config.add_request_method` adds extra methods to the request.

**Comment by domenkozar** [comment]
1. Yes, those could be defined as `Any` until zope.interface plugin exists.

2. Could define http://mypy.readthedocs.io/en/latest/protocols.html for default Configurator, then the rest of custom methods typing is up to the user.

**Comment by pior** [comment]
Is there a plan to work on typing already?

I don't see any mention of it in the 2.0 feature list #2362. 

I would love to help with stubs/comments for 1.x since I'm using it at work, but it would also be interesting to plan for Py3.

**Comment by ztane** [comment]
Please when talking with pycharm folks, tell them we want to type with zope interfaces too :D

**Comment by mmerickel** [comment]
Here is, in my eyes, what needs to happen to get typing into Pyramid. The goal would be to support typing via annotations and not via stub files or comment syntax. This is open for debate but this is the roadmap I have in my head.

- [ ] refactor the repo structure first #3022
- [ ] drop support for Python 2 #2903
- [ ] experiment / come up with a strategy to handle the configurator and z.i issues (using Any or something else) and test it out on pycharm and mypy
- [ ] actual PR to pyramid to add typings including mypy steps to our `tox -e lint` config

**Comment by jtrakk** [comment]
> Yes, those could be defined as Any until zope.interface plugin exists.

Here it is. https://github.com/Shoobx/mypy-zope

**Comment by duynhaaa** [comment]
Hi, has anyone started working on this? If not, I'm willing to contribute, but definitely need some guidance from the maintainers. I really love Pyramid's design and architect and hate seeing it slowly becoming a relic of the past.

**Comment by mmerickel** [comment]
I'm happy to review as best as I can - I just have limited familiarity with Python's type annotation system and haven't used things like mypy yet.

**Comment by luhn** [comment]
imo the best place to start would be with [webob](https://github.com/Pylons/webob), since `pyramid.request.Request` is the object that's most interacted with it and a large chunk of it is inherited from `webob.Request`.

**Comment by luhn** [comment]
For Pyramid itself, the biggest outstanding question to me is how to handle Configurator, especially `add_directive` and `add_request_method`.  Those are making runtime modifications to the Configurator and Request, respectively, and I'm not sure how to get that information to mypy.

**Comment by mmerickel** [comment]
You're correct that `add_directive` and `add_request_method` are the true challenges of typing pyramid. My general impression of how these would need to be handled is with external type extesions. Webob and Pyramid would maintain their specific types for all builtin behaviors on `pyramid.config.Configurator` and `webob.request.Request` and `pyramid.request.Request`. User code would then default to using these types. However we would need a way for a package to provide a type extension alongside the directive and the user code would need a way to use those types within their own code for type checking. Maybe there would be a way to then automate pulling those types together for mypy, but it starts with the basics:

- You define a directive, you provide a type extension to `pyramid.config.Configurator`.
- You define a request method, you provide a type extension to `pyramid.request.Request`.

Later a user could do something like:

```python
class ConfiguratorWithExtensionA(Configurator):
    def extension_a(...) -> None:
        pass

class MyConfigurator(Configurator, ConfiguratorWithExtensionA):
    pass

config: MyConfigurator = pyramid.config.Configurator()
with config:
    config.add_directive(..., 'extension_a')

    config.extension_a(...)

app = config.make_wsgi_app()
```

Again I'm showing how naive I am with Python types but I think the general strategy makes sense from what I've seen.

**Comment by merwok** [comment]
I think this may not work, as the analysis would find that `Configurator()` does not return a MyConfigurator instance, but there is `typing.cast` to override analysis.

**Comment by duynhaaa** [comment]
I think at the moment, the type checking of such dynamic objects (whose interfaces can be modified imperatively after the initial declaration, like the `pyramid.config.Configurator` with `.add_directive` as mentioned) to the users because:

1. Static type checking only works best on information that has already been statically and declaratively encoded in the source code (e.g., declarations of classes, functions and variables). The only dynamic thing that the type checking system can support is generic type, which requires all the possible derived types be declared, either explicitly or implicitly (if you squint your eyes enough). Think of it like conditional branching but for types.

2. Other languages with statically typed nature, such as Go and C/C++, also have to enter _"type-unsafe"_ zone at some point because it doesn’t make sense or worth the effort to perform _"type-gymnastic"_ just to make the only 5% unsafe code safe when tests are sufficient enough. Such cases usually happen when either the data is unstructured or the nature of the problem is too hard to define as a whole.

So in conclusion, unless we know all the possible future combination, which basically is omniscience at this point, or willing to introduce a complicated type system, which usually is not a good idea (mandatory example: https://github.com/gruhn/typescript-sudoku), there’s no need to have the framework fully typed to a tee.

TL;DR: Add typings for things we know and leave the unknown to the users.

**Comment by mmerickel** [comment]
Yes absolutely the initial goal is just to get the core framework to have better types. The extensibility of add_directive and add_request_method can come later. 

A lot of core types in the framework are already defined as zope interfaces in pyramid.interfaces so ideally we’d have a way to reuse that but I don’t think z.i is expressive enough to really do what you want with a protocol in mypy.

**Comment by duynhaaa** [comment]
So, should I start with `webob` like @luhn suggested or go straight to `pyramid`?

I will try to get something working soon (expecting by the end of next week).

**Comment by mmerickel** [comment]
I’d probably suggest starting with pyramid simple because webob’s codebase is in a weird cleanup phase that needs some love so I think it’d introduce annoying stop energy to start there.

**Comment by duynhaaa** [comment]
@mmerickel Is `webob` in the middle of a rewrite or something?

Either way, I'll start working on `pyramid`.

**Comment by mmerickel** [comment]
No not a rewrite. I looked at it again and it's not as bad as I thought, things are in a pretty decent state atm. You could definitely add types to webob. MultiDict will be tricky but hey :-)

Ultimately it'd be helpful because you'll want to inherit those types into pyramid.request.Request and pyramid.response.Response.

**Comment by mmerickel** [comment]
> I think this may not work, as the analysis would find that Configurator() does not return a MyConfigurator instance, but there is typing.cast to override analysis.

Good point yeah I suppose a cast would be required. It feels like there's not going to be a way around that but most of my typing experience is in static typing domains (or in the overly flexible z.i per-object typing domain which won't work with python's annotations afaik).

**Comment by merwok** [comment]
IIUC Python’s typing is static, but smart enough to support duck typing (with Protocols).
But the issue of adding methods is another thing.

**Comment by Daverball** [comment]
@duynhaaa I have some partially complete stubs for pyramid that still cover a decent amount of ground, I could send your way. From experience I can tell you that one set of type hints will be difficult to integrate into the current architecture where things like `Configurator` and `Request` are dynamically extended with additional properties and methods. This is also reflected in the stubs I have written, so they include properties and methods that don't actually exist on the base implementation, but are there for our specific choice of extensions.

The other large issue is the use of `zope.interface`, which is not very friendly towards type checker use, due to interfaces not being proper types and omitting the `self` argument on methods. There is a mypy plugin, but that means everyone will be locked into using mypy. I tried to start a discussion about those and other related problems here: https://github.com/Shoobx/mypy-zope/issues/122

In both cases it will probably be difficult to provide one set of type hints that will be useful to users, so I would actually avoid adding a `py.typed` marker file to `pyramid`. The best you can hope for currently, without large refactors, are type hints that are useful for checking the internal API. Then you could write tools so people can auto-generate stubs for their specific pyramid configuration based on those internal type hints and the project's configuration.

**Comment by sruggier** [comment]
Given that [Python 2 support has been dropped](https://github.com/Pylons/pyramid/issues/3653#issuecomment-1656544377), the current most straightforward option 1 would be to use Python 3's type annotation syntax directly in the source tree, right?

**Comment by tseaver** [comment]
>  the current most straightforward option 1 would be to use Python 3's type annotation syntax directly in the source tree, right?

I'm -1 on inlining the type info myself:  I've lived with it in large codebases for years now, and find the damage it does to readability / comprehensibility is ginormous.

**Comment by sruggier** [comment]
> I'm -1 on inlining the type info myself: I've lived with it in large codebases for years now, and find the damage it does to readability / comprehensibility is ginormous.

For what it's worth, the [Typing Python Libraries](https://typing.python.org/en/latest/guides/libraries.html#how-to-provide-type-annotations) page (in Python's typing documentation) has this recommendation:

> We recommend using the inline type annotations approach, since it has the following benefits:
> * Typically requires the least effort to add and maintain
> * Users don’t have to download additional packages
> * Always remains consistent with the implementation
> * Allows library authors to type check their own code
> * Allows language servers to show users relevant details about the implementation, such as docstrings and default parameter values

I find the main impact to readability comes from trying to combine multiple annotated items on the same line. One can efficiently avoid that by adding a trailing comma in places where it happens, like function arguments, so they'll be wrapped on save (after configuring one's editor to run `ruff format` on save).

**Comment by jvanasco** [comment]
> I'm -1 on inlining the type info myself: I've lived with it in large codebases for years now, and find the damage it does to readability / comprehensibility is ginormous.

I am +1 on this.  I find it incredibly useful.  I've also encountered far too many issues with codebases where the typing and docstrings are not aligned.  One example of this is Certbot, which changed many functions to `bytes` only from `bytes or str` – but did not update their docstrings.  API docs / source code says one thing, typing says another.  This can be deciphered easily with inline typing; it's a nightmare using stubs.

**Comment by stevepiercy** [comment]
@ericof can you jump in to describe the work that you plan to do with type annotations for Zope and Plone? It seems to me that since both Pyramid and Plone depend on Zope packages, that the implementations of type annotations across all three should align with one another.

**Comment by Daverball** [comment]
> Can we just define protocols next to each interface and then annotate APIs using those protocols? Quite a lot of boilerplate but seems doable. The z.i interfaces aren't as expressive and so it seems to me we just need to bite the bullet and define those protocols.

Absolutely, that's what I want to be able to automate. It won't be as type-safe as `zope.interface` aims to be, but pyramid doesn't really appear to be using interfaces in that strict sense anyways, since you often don't need to mark your implementations with `@implementer` in order for them to work in the `Configurator` methods that accept them.

The main drawback of going that route would be, that if folks were already using some of pyramid's interfaces in their type annotations, they would now need to change them all, but that seems less painful than forcing everyone to use `mypy` and `mypy-zope`, especially as more performant alternatives like `ty` and `pyrefly` become feature complete and stable enough to use.

Also regardless of whether you directly reference the interfaces or not, they will cause errors due to the missing `self` parameter, which will have to be silenced. So even if you give people alternatives, `zope.interface` still effectively becomes a nuisance if you're not using `mypy-zope`, so it might be a good idea to start there and update `zope.interface` to provide the same functionality in a way that's more compatible with type checkers out of the box and leave the old way of defining an interface as a compatibility shim for folks that don't want to update their code just to satisfy type checkers.
